compiler/rustc_codegen_llvm/src/debuginfo/mod.rs

```diff
@@ -438,12 +438,11 @@ impl DebugInfoMethods<'tcx> for CodegenCx<'ll, 'tcx> {
             substs: SubstsRef<'tcx>,
             name_to_append_suffix_to: &mut String,
         ) -> &'ll DIArray {
-            let type_name_cache = &mut *debug_context(cx).type_name_cache.borrow_mut();
```

Comment:
I think holding the RefCell "lock" here for longer than just the `type_names::push_generic_params()` call might have been the reason why you saw the BorrowMut panics because the `type_metadata()` call farther below will also try to lock the cache.